src/thread.rs

```diff
@@ -104,10 +114,182 @@ enum ThreadJoinStatus {
     Joined,
 }
 
+/// Set of sync objects that can have properties queried.
+/// The futex is not included since it can only signal
+/// and awake values.
+#[derive(Copy, Clone, PartialEq, Eq, Hash, Debug)]
+enum SyncObject {
+    /// Can query lock state.
+    Mutex(MutexId),
+    /// Can query lock state.
+    RwLock(RwLockId),
+    /// Can query the if awaited.
+    Condvar(CondvarId),
+
+}
+
+enum YieldRecordState {
+
+    /// The thread has made progress and so the recording of
+    /// states has terminated.
+    /// A thread is considered to have made progress performs some
+    /// action that is visible to another thread.
+    /// Examples include a release-store, mutex-unlock,
+    /// thread-termination and release-fence.
+    MadeProgress,
+
+    /// The thread is currently recording the variable watch
+    /// state for the yield operation, this watch operation
+    /// can occur once of multiple times to validate the yield
+    /// is correct.
+    /// Multiple iterations can prevent reporting live-lock
+    /// in yield loops with finite iterations
+    Recording {
+        /// Map of global memory state to an record index.
+        /// If set to zero then this location is not watched,
+        /// a value of 1 means that it was watched on the first
+        /// iteration and a value of x>=2 means that the value
+        /// was watched on the 1,2,..,x iterations.
+        watch_atomic: FxHashMap<AllocId, RangeMap<u32>>,
+
+        /// Map of synchronization objects that can have properties
+        /// queried and are currently watched by this thread.
+        watch_sync: FxHashMap<SyncObject, u32>,
+        
+        /// The current iteration of the yield livelock loop
+        /// recording, should always be less than or equal
+        /// to the global live-lock loop counter.
+        record_iteration: u32,
+    }
+}
+
+impl YieldRecordState {
```

Comment:
Looks like this type could potentially live in a different file? Would that make sense, to separate the yield-watching logic from the rest of the scheduler?